core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java

```diff
@@ -777,19 +777,101 @@ private boolean hasDescClusteringOrder(TableMetadata metadata) {
   public void createIndex(
       String namespace, String table, String columnName, Map<String, String> options)
       throws ExecutionException {
+
     try (Connection connection = dataSource.getConnection()) {
+      alterToIndexColumnTypeIfNecessary(connection, namespace, table, columnName);
       createIndex(connection, namespace, table, columnName);
       updateTableMetadata(connection, namespace, table, columnName, true);
-    } catch (SQLException e) {
+    } catch (ExecutionException | SQLException e) {
       throw new ExecutionException("creating the secondary index failed", e);
     }
   }
 
+  private void alterToIndexColumnTypeIfNecessary(
+      Connection connection, String namespace, String table, String columnName)
+      throws ExecutionException, SQLException {
+    DataType indexType = getTableMetadata(namespace, table).getColumnDataType(columnName);
+    ImmutableMap<DataType, String> typeMappingForIndexes = DATA_TYPE_MAPPING_FOR_KEY.get(rdbEngine);
+    assert typeMappingForIndexes != null;
+
+    String indexedColumnType = typeMappingForIndexes.get(indexType);
+    if (indexedColumnType == null) {
+      // The column type does not need to be altered to be compatible with being secondary index
+      return;
+    }
+    String alterColumnStatement =
+        getAlterColumnTypeStatement(namespace, table, columnName, indexedColumnType);
```

Comment:
I feel that triggering alter table by creating an index might be a bit too tricky?
I think it is safer to make users decide instead of changing schema automatically.

Reply:
@Torch3333 Sorry for the late feedback.
I discussed it with Toshi and now think it's OK to add.